src/Nethermind/Nethermind.Merge.Plugin/Synchronization/PivotUpdator.cs

```diff
@@ -126,38 +126,29 @@ private async void OnSyncModeChanged(object? sender, SyncModeChangedEventArgs sy
 
     private async Task<bool> TrySetFreshPivot(CancellationToken cancellationToken)
     {
-        Hash256? finalizedBlockHash = TryGetFinalizedBlockHashFromCl();
+        (Hash256 Hash, long Number)? potentialPivotData = await TryGetPivotData(cancellationToken);
 
-        if (finalizedBlockHash is null || finalizedBlockHash == Keccak.Zero)
-        {
-            return false;
-        }
-
-        long? finalizedBlockNumber = TryGetFinalizedBlockNumberFromBlockCache(finalizedBlockHash);
-        finalizedBlockNumber ??= TryGetFinalizedBlockNumberFromBlockTree(finalizedBlockHash);
-        finalizedBlockNumber ??= await TryGetFinalizedBlockNumberFromPeers(finalizedBlockHash, cancellationToken);
-
-        return finalizedBlockNumber is not null && TryOverwritePivot(finalizedBlockHash, (long)finalizedBlockNumber);
+        return potentialPivotData is not null && TryOverwritePivot(potentialPivotData.Value.Hash, potentialPivotData.Value.Number);
     }
 
-    private Hash256? TryGetFinalizedBlockHashFromCl()
+    protected virtual async Task<(Hash256 Hash, long Number)?> TryGetPivotData(CancellationToken cancellationToken)
     {
+        // getting finalized block hash as it is safe, because can't be reorganized
         Hash256? finalizedBlockHash = _beaconSyncStrategy.GetFinalizedHash();
 
         if (finalizedBlockHash is null || finalizedBlockHash == Keccak.Zero)
         {
-            if (_logger.IsInfo && (_maxAttempts - _attemptsLeft) % 10 == 0) _logger.Info($"Waiting for Forkchoice message from Consensus Layer to set fresh pivot block [{_maxAttempts - _attemptsLeft}s]");
-
+            PrintWaitingForMessageFromCl();
```

Comment:
when you invert this `if` then the override looks quite similar in structure and probably can be mostly shared.